ldms/src/ldmsd/ldmsd_stream.c

```diff
@@ -317,68 +431,26 @@ int ldmsd_stream_publish_file(const char *stream, const char *type,
 		msglog("Error %d connecting to remote peer\n", conn_status);
 		return conn_status;
 	}
-	size_t max_msg = ldms_xprt_msg_max(x);
-	ldmsd_req_hdr_t req = malloc(max_msg);
-	if (!req) {
-		msglog("Error allocating %d bytes of memory for buffer\n",
-			max_msg);
-		return ENOMEM;
-	}
 
-	/* Put the instance name at the front of the request */
-	attr = (ldmsd_req_attr_t)(req + 1);
-	attr->discrim = 1;
-	attr->attr_id = LDMSD_ATTR_NAME;
-	attr->attr_len = strlen(stream) + 1;
-	size_t meta =
-		sizeof(*req) +
-		sizeof(*attr) + attr->attr_len + /* plugin name */
-		sizeof(*attr) +			 /* plugin data */
-		sizeof(attr->discrim);		 /* terminator */
-	strcpy((char *)attr->attr_value, stream);
-	first_attr = ldmsd_next_attr(attr);
-	ldmsd_hton_req_attr(attr);
-
-	uint32_t flags = LDMSD_REQ_SOM_F;
-	while ((data_len = fread(buffer, 1, sizeof(buffer), file)) > 0) {
-		if (conn_status)
-			return conn_status;
-
-		data_ptr = buffer;
-		while (data_len) {
-			this_rec = meta + data_len;
-			this_rec = this_rec < max_msg ? this_rec : max_msg;
-
-			attr = first_attr;
-			attr->discrim = 1;
-			attr->attr_id = attr_id;
-			attr->attr_len = this_rec - meta;
-			memcpy(attr->attr_value, data_ptr, attr->attr_len);
-			data_ptr += attr->attr_len;
-			data_len -= attr->attr_len;
-
-			next_attr = ldmsd_next_attr(attr);
-			ldmsd_hton_req_attr(attr);
-			next_attr->discrim = 0;
-
-			req->marker = LDMSD_RECORD_MARKER;
-			req->type = LDMSD_REQ_TYPE_CONFIG_CMD;
-			req->req_id = LDMSD_STREAM_PUBLISH_REQ;
-			req->rec_len = this_rec;
-			if (!data_len)
-				/* No more data, turn on last message flag */
-				flags |= LDMSD_REQ_EOM_F;
-			req->flags = flags;
-			ldmsd_hton_req_hdr(req);
-
-			rc = ldms_xprt_send(x, (char *)req, this_rec);
+	if (stream_type == LDMSD_STREAM_JSON) {
+		/* must send entire file at once */
+		char *_buf = fread_all(file, &data_len);
+		rc = ldmsd_stream_publish(x, stream, LDMSD_STREAM_JSON, _buf, data_len);
+		free(_buf);
+	} else {
+		/* can send stream in chunks */
+		while ((data_len = fread(buffer, 1, sizeof(buffer), file)) > 0) {
+			rc = ldmsd_stream_publish(x, stream, LDMSD_STREAM_STRING,
+					buffer, data_len);
 			if (rc)
 				goto err;
-			/* Turn off start of message flag, for subsequent sends */
-			flags = 0;
 		}
 	}
 
+	while (ldms_xprt_sq_status(x) == EBUSY) {
```

Comment:
This can't be correct right? If this is required, then the transport is broken.

Reply:
It is "broken" because we don't have send completion. For a big data case, multiple records may still be in the io_q (the pending io) or sq (in sock, not submitted to the system yet). Since it is asynchronous, we may close the xprt before all records are submitted to the underlying transport. And I found that from my test last night.

Reply:
@tom95858 shall we pull in `send_mapped()` from master instead? I believe `send_mapped()` has send completion and we can use it for the case like this.

Reply:
I remember last you guys address the send without completion problem for stream, we have an assumption that users will use stream to send small data. That's why we didn't put in the zap send_mapped() at the time or made a bigger patch.

Reply:
The problem is that the close is synchronous, but the send is asynchronous. IOW, we queue sends to a send_q, but we don't queue close. We had this same issue on receive a while back, i.e. the inbound disconnect would get delivered right away which could then be followed by i/o completions. 

The other issue is that the chunking of configuration messages should be done in one place not be copied here and there so that it has to be maintained in multiple places.

It's to clear to me why configuration messages should be handled differently than any other request, e.g. dir_reply, that may be very large.

Reply:
I agree with you. The config was done outside of ldms xprt (i.e. ldmsd) because historically, according to @nichamon, the config was only used by ldmsd and at the time no other LDMS was doing splitting/joining messages like `push` and `dir` do today.

I think queuing `close` command is a fine, but we still need a send completion for `ldmsd_stream_publish` program, because the program may exited before all messages got submitted (the routine of the program was connect, repeatedly send, close, exit).

This patch was a hurried/last-minute "make it work" kind of effort. I was actually thinking something similar to you but the change will be much bigger and more chance of disruption. Shall I or @nichamon started a proposal and we go from there?

I'm closing this pull request in favor of the better solution in a near future.